discojs/src/privacy.ts

```diff
@@ -12,18 +9,19 @@ import { tf } from '.'
  * @param task the task
  * @returns the noised weights for the current round
  */
-export function addDifferentialPrivacy (updatedWeights: Weights, staleWeights: Weights, task: Task): Weights {
+export function addDifferentialPrivacy (updatedWeights: WeightsContainer, staleWeights: WeightsContainer, task: Task): WeightsContainer {
   const noiseScale = task.trainingInformation?.noiseScale
   const clippingRadius = task.trainingInformation?.clippingRadius
 
-  const weightsDiff = List(updatedWeights)
-    .zip(List(staleWeights))
-    .map(([w1, w2]) => w1.sub(w2))
-  let newWeightsDiff: List<tf.Tensor>
+  const weightsDiff = updatedWeights.sub(staleWeights)
+  let newWeightsDiff: WeightsContainer
 
   if (clippingRadius !== undefined) {
     // Frobenius norm
-    const norm = Math.sqrt(weightsDiff.map((w) => w.square().sum().dataSync()[0]).reduce((a: number, b) => a + b))
+    const norm = Math.sqrt(
```

Comment:
don't we have a difference and a norm function in the weights helper file?

Reply:
or does this here avoid one copy or so? (weights are large so should avoid memory duplications etc)

Reply:
We don't have have any utility for the Frobenius norm, so I compute it manually from the weights container's tensors.  We already have an `add` method for adding weights containers together entry-wise, but the `reduce` here adds all entries of a single weights container together.

I moved `dataSync()` outside of the first map to avoid unnecessarily transferring tensors from GPU to CPU for each entry of the weights matrix. Now, `dataSync` is called after the reduction to a single scalar, which is much more efficient.